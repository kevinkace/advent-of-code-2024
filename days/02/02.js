export function testSafe(data) {
    let dir;

    return data.every((currValue, index, arr) => {
        const nextValue = arr[index + 1];

        // done! (watchout for 0)
        if (nextValue === undefined) {
            return true;
        }

        if (nextValue === currValue) {
            return false;
        }

        // first iteration, get direction
        if (!index) {
            dir = nextValue > currValue ? 1 : -1;
        }

        if (
            nextValue > (currValue * dir) &&
            Math.abs(nextValue - currValue) <= 3
        ) {
            return true;
        }

        return false;
    });
}

function safeValues(currValue, nextValue, dir) {
    // done! (watchout for 0)
    if (nextValue === undefined) {
        return true;
    }

    if (nextValue === currValue) {
        return false;
    }

    const currDir = nextValue > currValue ? 1 : -1;

    if (dir && currDir !== dir) {
        return false;
    }

    if (
        nextValue > (currValue * currDir) &&
        Math.abs(nextValue - currValue) <= 3
    ) {
        return currDir;
    }

    return false;
}

export function testSafeTolerance(data, tolerance = 1) {
    let dir;
    let faults = 0;

    for (let i = 0; i < data.length - 1; i++) {
        for (let j = i+1; j < data.length; j++) {
            let nextI = i + j;
            let currDir;

            currDir = safeValues(data[i], data[nextI], currDir);
            console.log({currDir})

            if (
                currDir === false ||
                dir && dir !== currDir
            ) {
                faults++;

                continue;
            }

            if (currDir === true) {
                break;
            }

            if (currDir === true) {
                return faults <= tolerance;
            }

            dir === dir || currDir;
        }

        console.log("end of outer loop")
    }

    return false;
}

export function validateDiffs(diffs) {
    let dir;

    return diffs.every(diff => {
        const currDir = diff > 0 ? 1 : -1;

        if (dir && dir !== currDir || diff === 0) {
            return false;
        }

        dir = currDir;

        return Math.abs(diff) <= 3;
    });
}

export function validateDiffsTolerance(diffs, tolerance = 1) {
    let dir;
    let faults = 0;


    // i is not incremented here!
    for (let i = 0; i < diffs.length; /* NOTHING */) {
        const currDiff = diffs[i];
        const currDir = currDiff > 0 ? 1 : -1;

        const invalid = dir && dir !== currDir || currDiff === 0 || Math.abs(currDiff) > 3;

        if (invalid) {
            faults++;

            diffs.splice(i, 1);
            diffs[i] = currDiff + diffs[i];

        } else {
            i++;
        }

        if (faults > tolerance) {
            return false;
        }

        dir = dir || currDir;
    }

    return true;
}

function getDir(diff) {
    return diff > 0 ? 1 : -1;
}

export function checkDirection(diffs) {
    if (diffs.join().includes("0")) {
        return false;
    }

    let dir = getDir(diffs[0]);

    return diffs.every(diff => (dir === getDir(diff)));
}

export function checkDistances(diffs, { max = 3 } = {}) {
    return diffs.every(diff => Math.abs(diff) <= max);
}


export function valDiffs2099(diffs) {
    return checkDirections(diffs) && checkDistances(diffs);
}