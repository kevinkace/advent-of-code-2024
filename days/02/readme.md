# Day {day}

https://adventofcode.com/2024/day/2



## Part 1

Simple parsing. My initial implementation had 2 condition checks: one for positive and one for negative direction. Another option is to use the direction as a multiplier when doing comparisons.

```js
// separate positive and negative checks
if (
    dir === 1 &&
    nextValue > (currValue * dir) &&
    nextValue - currValue <= 3
) {
    return true;
}

if (
    dir === -1 &&
    nextValue < currValue &&
    currValue - nextValue <= 3
) {
    return true;
}

// cooler single check
if (
    nextValue > (currValue * dir) &&
    Math.abs(nextValue - currValue) <= 3
) {
    return true;
}
```

I actually prefer the first option, it reads clearer. But this is AOC so some code golfing is the point, and often comes in handy later on as computational demands increase.

## Part 2

being able to isolate tests is very important.


## Closing


