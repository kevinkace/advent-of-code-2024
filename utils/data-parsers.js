const EOL = "\n";
const DELIM = "   ";

export function splitToLists(data, { eol = EOL, delim = DELIM } = {}) {
    return data.split(eol).map(line => line.split(delim)).reduce((acc, cur) => {
        acc[0].push(Number(cur[0]));
        acc[1].push(Number(cur[1]));

        return acc;
    }, [[], []]);
}
