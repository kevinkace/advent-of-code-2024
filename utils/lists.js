export function getListDiffs([listA, listB]) {
    const diffs = [];

    for (let index = 0; index < listA.length; index++) {
        diffs.push(Math.abs(listA[index] - listB[index]));
    }

    return diffs;
}

export function sortLists(lists) {
    return lists.map(list => list.sort((a, b) => a > b ? 1 : -1));
}

export function sumList(list) {
    return list.reduce((acc, cur) => acc + cur, 0);
}

export function tally(list) {
    return list.reduce((acc, cur) => {
        const newVal = (acc.get(cur) || 0) + 1;

        acc.set(cur, newVal);

        return acc;
    }, new Map());
}

export function tallyProduct(tallyA, tallyB) {
    const products = [];

    const [shorterTally, longerTally] = [tallyA, tallyB].sort((tA, tB) => tA.length < tB.length ? 1 : -1);

    shorterTally.forEach((sTallyOccurrence, number) => {
        const lTallyOccurrence = longerTally.get(number) || 0;

        products.push(sTallyOccurrence * lTallyOccurrence * number);
    });

    return sumList(products);
}

export function arrToDiffs(arr) {
    return arr.reduce((acc, curr, index) => {
        const next = arr[index + 1];

        if (next === undefined) {
            return acc;
        }

        acc.push(next - curr);

        return acc;
    }, []);
}
