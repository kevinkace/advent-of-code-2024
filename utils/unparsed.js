
const EOL = "\r\n";

export function letterInbounds(string, { row, col }) {
    const rows = string.split(EOL);

    return row >= 0 && row < rows.length && col >= 0 && col < rows[0].length;
}

export function letterAt(string, { row, col }) {
    if (!letterInbounds(string, { row, col })) {
        return;
    }

    const rows = string.split(EOL);

    return rows[row][col];
}

export function numberRows(string) {
    return string.split(EOL)[0].length;
}
