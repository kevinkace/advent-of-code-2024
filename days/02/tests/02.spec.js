import { describe, expect, test } from "vitest";

import { parse2d } from "../../../utils/data-parsers";

import example1 from "../data/example-1";
import part1 from "../data/02-1";

import { testSafe, testSafeTolerance, validateDiffs, validateDiffsTolerance } from "../02";
import { arrToDiffs } from "../../../utils/lists";

const parsedExample = parse2d(example1);
const parsedData = parse2d(part1);

describe("day 02 - 1", () => {
    test("example", () => {
        expect(parsedExample).toMatchSnapshot();

        expect(testSafe(parsedExample[0])).toBe(true);

        expect(parsedExample.map(testSafe)).toMatchSnapshot();
    });

    test("ad hoc", () => {
        const rows = parse2d("21 20 19 16 15 14 11 9");

        expect(testSafe(rows[0])).toBe(true);
    });

    test("part 1", () => {
        const trueCount = parsedData.map(testSafe).filter(Boolean).length;

        console.log({ trueCount });

        expect(trueCount).toBeGreaterThan(512);
    });
});

describe("day 02 - 2", () => {
    test.skip("example", () => {
        const trueCount = parsedExample.map(row => testSafeTolerance(row)).filter(Boolean).length;

        expect(trueCount).toBe(4);
    });

    test("ad hoc - single row", () => {
        const diffs4 = arrToDiffs(parsedExample[4]);

        expect(diffs4).toMatchSnapshot();
    });

    test.only("ad hoc - pop first", () => {
        const data = parsedData[24];
        const diffs = arrToDiffs(data);

        console.log({ data, diffs });
        const valid = validateDiffsTolerance(diffs);

        console.log({ valid, diffs, data });
    });

    test("ad hoc", () => {


        // expect(validateDiffs(diffs)).toBe(false);

        // const allDiffs = parsedExample.map(row => validateDiffs(arrToDiffs(row)));

        // expect(allDiffs).toMatchSnapshot();


        // expect(validateDiffsTolerance(arrToDiffs(parsedExample[3]))).toBe(true);

        const allValidations = parsedExample.map(row => {
            const diffs = arrToDiffs(row);

            return validateDiffsTolerance(diffs);
        });

        expect(allValidations).toMatchSnapshot();
    });

    test("part 2", () => {
        const allValidations = parsedData.map(row => {
            const diffs = arrToDiffs(row);

            return validateDiffsTolerance(diffs);
        });

        const trueCount = allValidations.filter(Boolean).length;

        console.log({ trueCount });

        console.table(allValidations.map((v, i) => {
            if (!v) {
                return { i, d : parsedData[i].join(","), diffs : arrToDiffs(parsedData[i]).join(",") };
            }
        }).filter(Boolean));

        expect(trueCount).toBeGreaterThan(592);
        expect(trueCount).toBeLessThan(800);

        // expect(allValidations).toEqual();
    });
});
