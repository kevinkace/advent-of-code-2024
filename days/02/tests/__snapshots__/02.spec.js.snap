// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`day 02 - 1 > example 1`] = `
[
  [
    7,
    6,
    4,
    2,
    1,
  ],
  [
    1,
    2,
    7,
    8,
    9,
  ],
  [
    9,
    7,
    6,
    2,
    1,
  ],
  [
    1,
    3,
    2,
    4,
    5,
  ],
  [
    8,
    6,
    4,
    4,
    1,
  ],
  [
    1,
    3,
    6,
    7,
    9,
  ],
]
`;

exports[`day 02 - 1 > example 2`] = `
[
  true,
  false,
  false,
  false,
  false,
  true,
]
`;

exports[`day 02 - 2 > ad hoc - single row 1`] = `
[
  -2,
  -2,
  0,
  -3,
]
`;

exports[`day 02 - 2 > ad hoc 1`] = `
[
  true,
  false,
  false,
  true,
  true,
  true,
]
`;
